Support Bonsai assets versions prefixed with v

palourde opened the pull request:
Signed-off-by: Simon Plourde <simon@sensu.io>

## What is this change?

It adds supports for Bonsai assets that use the version format `vX.Y.Z`, in addition to semver (`X.Y.Z`).

## Why is this change necessary?

Closes https://github.com/sensu/sensu-go/issues/3425

## Does your change need a Changelog entry?

Added

## Do you need clarification on anything?

I tried my best to make sure I didn't forget anything; it appears this package is only used in sensuctl but I might be forgetting something.

## Were there any complications while making this change?

Nope

## Have you reviewed and updated the documentation for this change? Is new documentation required?

Nope

## How did you verify this change?

Manually tested it and added a unit test where available.

## Is this change a patch?

Yep

julian7 commented:
The only problem with using `version.Original` is that it is not the version which is equivalent to the version the bonsai asset describes, but it must be exactly in the same format as it is advertised.

Example: if a bonsai asset has v0.0.1 and v0.0.2, but the version "0.0.2" has been specified, there will be a match (since converted "v0.0.2" and "0.0.2" are equivalent), but then it will fail at pulling the selected version. One has to specify "v0.0.2", the exact version string, to download.

I think we should pull the exact string from the version list, if a match is found, and it should use its original to do downloads.

palourde commented:
@julian7 I pushed a new commit which address this concern:

```sh
$ sensuctl asset add nixwiz/sensu-go-fatigue-check-filter:v0.2.2
fetching bonsai asset: nixwiz/sensu-go-fatigue-check-filter:v0.2.2
added asset: nixwiz/sensu-go-fatigue-check-filter:v0.2.2
$ sensuctl asset add nixwiz/sensu-go-fatigue-check-filter:0.2.2
fetching bonsai asset: nixwiz/sensu-go-fatigue-check-filter:v0.2.2
added asset: nixwiz/sensu-go-fatigue-check-filter:v0.2.2
$ sensuctl asset outdated
               Asset Name                            Bonsai Asset               Current Version   Latest Version
 ────────────────────────────────────── ────────────────────────────────────── ───────────────── ────────────────
  nixwiz/sensu-go-fatigue-check-filter   nixwiz/sensu-go-fatigue-check-filter   v0.2.2            0.4.0
```

palourde commented:
Here's the expected behavior now:

When installing an asset

| Command        | Bonsai Versions           | Result  |
| ------------- |:-------------:| -----:|
| `sensuctl asset add sensu/foo`     | `0.1.0`, `v0.2.0` | Adding `v0.2.0` |
| `sensuctl asset add sensu/foo:v0.1.0`     | `0.1.0`, `v0.2.0` | Adding `0.1.0` |
| `sensuctl asset add sensu/foo:0.2.0`     | `0.1.0`, `v0.2.0` | Adding `v0.2.0` |
| `sensuctl asset outdated`     | `0.1.0`, `v0.2.0` | `0.1.0`is outdated |
| `sensuctl asset outdated`     | `0.1.0`, `v0.2.0`, `0.3.0` | `0.1.0` & `v0.2.0` are outdated |

julian7 commented:
I approve this change (whatever it means). It already made me smile yesterday, when I went through the code changes.

echlebek commented on bonsai/bonsai.go:
```diff
@@ -78,14 +93,46 @@ func NewBaseAsset(name string) (*BaseAsset, error) {
+			return nil, fmt.Errorf("asset %q does not have any available versions", b.Name)
+		}
+
+		fmt.Println("no version specified, using latest:", v.Original())
```
Should probably be removed?

palourde replied:
It was actually already present, which produces the following:

```
$ sensuctl asset add nixwiz/sensu-go-fatigue-check-filter
no version specified, using latest: 0.4.0
fetching bonsai asset: nixwiz/sensu-go-fatigue-check-filter:0.4.0
added asset: nixwiz/sensu-go-fatigue-check-filter:0.4.0
```

echlebek replied:
Or maybe moved to somewhere less library-ish?
